tests/pocket_test.cpp

```diff
@@ -0,0 +1,832 @@
+
+        THEN( "max_contains_volume is based on the pocket's ammo type" ) {
+            CHECK( data_ammo_box.max_contains_volume() == 1000_ml );
+        }
+    }
+}
+
+// Ammo restriction
+// ----------------
+// Pockets with "ammo_restriction" defined have capacity based on those restrictions (from the
+// type(s) of ammo and quantity of each. They can only contain ammo of those types and counts,
+// and mixing different ammos in the same pocket is prohibited.
+//
+// Related JSON fields:
+// "ammo_restriction"
+//
+// Functions:
+// item_pocket::can_contain
+// item_pocket::insert_item
+//
+TEST_CASE( "magazine with ammo restriction", "[pocket][magazine][ammo_restriction]" )
+{
+    pocket_data data_mag( item_pocket::pocket_type::MAGAZINE );
+
+    // ammo_restriction takes precedence over volume/length/weight,
+    // so it doesn't matter what these are set to - they can even be 0.
+    data_mag.volume_capacity = 0_liter;
+    data_mag.min_item_volume = 0_liter;
+    data_mag.max_item_length = 0_meter;
+    data_mag.max_contains_weight = 0_gram;
+
+    // TODO: Add tests for multiple ammo types in the same clip
+
+    GIVEN( "magazine has 9mm ammo_restriction" ) {
+        // Magazine ammo_restriction may be a single { ammotype, count } or a list of ammotypes and
+        // counts. This mag 10 rounds of 9mm, as if it were defined in JSON as:
+        //
+        //      "ammo_restriction": { "9mm", 10 }
+        //
+        const int full_clip_qty = 10;
+        data_mag.ammo_restriction.emplace( ammotype( "9mm" ), full_clip_qty );
+        item_pocket pocket_mag( &data_mag );
+
+        WHEN( "it does not already contain any ammo" ) {
+            REQUIRE( pocket_mag.empty() );
+
+            THEN( "it can contain a full clip of 9mm ammo" ) {
+                item ammo_9mm( "test_9mm_ammo", 0, full_clip_qty );
+                expect_can_contain( pocket_mag, ammo_9mm );
+            }
+
+            THEN( "it cannot contain items of the wrong ammo type" ) {
+                item rock( "test_rock" );
+                item ammo_45( "test_45_ammo", 0, 1 );
+                expect_cannot_contain( pocket_mag, rock, "item is not the correct ammo type",
+                                       item_pocket::contain_code::ERR_AMMO );
+                expect_cannot_contain( pocket_mag, ammo_45, "item is not the correct ammo type",
+                                       item_pocket::contain_code::ERR_AMMO );
+            }
+
+            THEN( "it cannot contain items that are not ammo" ) {
+                item rag( "test_rag" );
+                expect_cannot_contain( pocket_mag, rag, "item is not an ammo",
+                                       item_pocket::contain_code::ERR_AMMO );
+            }
+        }
+
+        WHEN( "it is partly full of ammo" ) {
+            int half_clip_qty = full_clip_qty / 2;
+            item ammo_9mm_half_clip( "test_9mm_ammo", 0, half_clip_qty );
+            pocket_mag.insert_item( ammo_9mm_half_clip );
```

Comment:
could be an excuse to use expect_can_insert here?

Reply:
Done.